internal/auth/providers/amazon_cognito.go

```diff
@@ -0,0 +1,395 @@
+// The code is redeemed for an access token and refresh token
+// 1. POSTs the code and grant_type to https://docs.aws.amazon.com/cognito/latest/developerguide/token-endpoint.html
+// 2. If the request fails, the authenticator will return a 500 and display an error page (see oauth_proxy.go#OAuthCallback)
+// 3. If the request succeeds, the data from AmazonCognito contains:
+//     - the access token which we use to get data from AmazonCognito
+//     - the refresh token which we can use to get a new access_token
+//     - the expiration time of the access token
+//     - a Base64 encoded id token which contains the user's email
+//       address and whether or not that email address is verified
+func (p *AmazonCognitoProvider) Redeem(redirectURL, code string) (*sessions.SessionState, error) {
+	logger := log.NewLogEntry()
+
+	if code == "" {
+		return nil, ErrBadRequest
+	}
+	params := url.Values{}
+	params.Add("redirect_uri", redirectURL)
+	params.Add("client_id", p.ClientID)
+	params.Add("code", code)
+	params.Add("grant_type", "authorization_code")
+	params.Add("scope", p.Scope)
+
+	var response struct {
+		AccessToken  string `json:"access_token"`
+		RefreshToken string `json:"refresh_token"`
+		ExpiresIn    int64  `json:"expires_in"`
+		IDToken      string `json:"id_token"`
+	}
+
+	err := p.amazonCognitoRequest("POST", p.RedeemURL.String(), params, []string{"action:redeem"}, nil, true, &response)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info(fmt.Sprintf("response %#v", response))
+
+	email, err := p.verifyEmailWithAccessToken(response.AccessToken)
+	if err != nil {
+		return nil, err
+	}
+
+	return &sessions.SessionState{
+		AccessToken:  response.AccessToken,
+		RefreshToken: response.RefreshToken,
+
+		RefreshDeadline:  sessions.ExtendDeadline(time.Duration(response.ExpiresIn) * time.Second),
+		LifetimeDeadline: sessions.ExtendDeadline(p.SessionLifetimeTTL),
+		Email:            email,
+	}, nil
+}
+
+// verifyEmailWithAccessToken takes in an access token and
+// checks if it is verified with AmazonCognito. Conditionally returns
+// an empty string, or if validated the email address in question.
+func (p *AmazonCognitoProvider) verifyEmailWithAccessToken(accessToken string) (string, error) {
+	if accessToken == "" {
+		return "", ErrBadRequest
+	}
+
+	userinfo, err := p.GetUserProfile(accessToken)
+	if err != nil {
+		return "", err
+	}
+
+	if userinfo.EmailAddress == "" {
+		return "", errors.New("missing email")
+	}
+
+	return userinfo.EmailAddress, nil
+}
+
+// ValidateGroupMembership takes in an email, a list of allowed groups an access token
+// - compares allowed groups to users' groups.
+// Conditionally returns an empty (nil) slice or a slice of the matching groups.
+func (p *AmazonCognitoProvider) ValidateGroupMembership(email string, allowedGroups []string, accessToken string) ([]string, error) {
+	// TODO: This is currently non-functioning
+	return allowedGroups, nil
+}
```

Comment:
I'm not sure what we'd want the behaviour to be in the case of no group membership functionality. At the moment this effectively means the group validation check always passes, right?